benches/search_comparison.rs

```diff
@@ -0,0 +1,276 @@
+            search_bench_case::<MAX, T, _>(
+                "sorted_vec",
+                make_sorted_vec,
+                search_sorted_vec,
+                &mut group,
+                i,
+                true,
+            );
+            search_bench_case::<MAX, T, _>(
+                "btreeset",
+                make_btreeset,
+                search_btreeset,
+                &mut group,
+                i,
+                true,
+            );
+            search_bench_case::<MAX, T, _>(
+                "ordsearch",
+                make_this,
+                search_this,
+                &mut group,
+                i,
+                true,
+            );
+        }
+        group.finish();
+    }
+
+    {
+        let groupname = format!("Construction {}", std::any::type_name::<T>());
+        let mut group = c.benchmark_group(groupname);
+        group.warm_up_time(WARM_UP_TIME).measurement_time(MEASUREMENT_TIME).plot_config(plot_config.clone());
+
+        for i in sizes.iter() {
+            construction_bench_case::<MAX, T, _>(
+                "sorted_vec",
+                make_sorted_vec,
+                &mut group,
+                i,
+                false,
+            );
+            construction_bench_case::<MAX, T, _>("btreeset", make_btreeset, &mut group, i, false);
+            construction_bench_case::<MAX, T, _>("ordsearch", make_this, &mut group, i, false);
+        }
+        group.finish();
+    }
+
+    {
+        let groupname = format!(
+            "Construction (with duplicates) {}",
+            std::any::type_name::<T>()
+        );
+        let mut group = c.benchmark_group(groupname);
+        group.warm_up_time(WARM_UP_TIME).measurement_time(MEASUREMENT_TIME).plot_config(plot_config);
+
+        for i in sizes.iter() {
+            construction_bench_case::<MAX, T, _>(
+                "sorted_vec",
+                make_sorted_vec,
+                &mut group,
+                i,
+                false,
+            );
+            construction_bench_case::<MAX, T, _>("btreeset", make_btreeset, &mut group, i, false);
+            construction_bench_case::<MAX, T, _>("ordsearch", make_this, &mut group, i, false);
+        }
+        group.finish();
+    }
+}
+
+fn search_bench_case<const MAX: usize, T, Coll>(
+    name: &str,
+    setup_fun: impl Fn(Vec<T>) -> Coll,
+    search_fun: impl Fn(&Coll, T) -> Option<&T>,
+    group: &mut BenchmarkGroup<WallTime>,
+    i: &usize,
+    duplicates: bool,
+) where
+    T: TryFrom<usize> + Ord + std::ops::Rem<Output = T> + num_traits::ops::wrapping::WrappingMul,
+    <T as TryFrom<usize>>::Error: core::fmt::Debug,
+{
+    group.bench_with_input(BenchmarkId::new(name, i), i, |b, i| {
+        let size = *i;
+        let v: Vec<T> = if duplicates {
+            (0..*i)
+                .map(|int| {
+                    let int = std::cmp::min(int, MAX);
+                    T::try_from(int / 16 * 16).unwrap()
+                })
```

Comment:
FYI to @Qqwy and @jonhoo. Here we are using saturated calculations, so for `u8` and MAX=65536 the resulting sequence will be – 0, 1, 2, ..., 254, 255, ..., 255, 255. Consider using wrap-around schematics (eg. 0, 1, 2, ..., 254, 255, 0, 1, 2, ..., 255, ...) so after sorting the payload collection will be more balanced.

Reply:
Oh, that's a very good point! We should change this one and the one in `construction` to instead use `int % (16 * 16)` :+1: